### DISCUSSION TITLE
How to assign a train_mask, test_mask and valid_mask for a PyG graph created from networkx

### DISCUSSION THREAD

**Discussion by RamachandranSitaraman-hv** [pyg-team/pytorch_geometric#4125] [Q&A] [answered]
Hi,
  How to assign a train_mask, test_mask and valid_mask for a PyG graph created from networkx...is there some property of the node i can set and have PyG automatically execute the masks while training, testing and validation? Thanks

**Comment by rusty1s** [comment, marked as answer]
You can add splitting masks automatically via `torch_geometric.transforms.RandomNodeSplit`. Let me know if this fixes your issues :)

**Reply by RamachandranSitaraman-hv** [reply]
Thanks Rusty1s,  is there a way i can control the splits by explicitly defining the nodes that go into training test and validation masks...?  Also is there a way givena PyG graph i can list the nodes, their IDs and their attributes?  Thanks

**Reply by rusty1s** [reply]
You can also define your own masks and simply add them to your `data` object, e.g.:
```python
train_mask = torch.zeros(data.num_nodes, dtype=torch.bool)
train_mask[torch.tensor([1, 2, 4, 6, ...]) = True  # The node indices used for training
```
Note sure what you mean with listing their IDs. PyG refers to nodes by indices from 0 to number of nodes - 1. As such, to get the node features of node 0, you can simply do `print(data.x[0])`.

**Reply by RamachandranSitaraman-hv** [reply]
Thanks for the response Rusty1s, once the train_mask is defined as you have indicated, what is the syntax to add it to the data object?  What you have mentioned listing the attributes by their ids is what i was looking for, thanks

**Reply by rusty1s** [reply]
Just assign it:
```python
data.train_mask = train_mask
```

**Reply by RamachandranSitaraman-hv** [reply]
okay great thanks

**Reply by RamachandranSitaraman-hv** [reply]
I take that that once i assign it, the rest is taken care of automatically by the API for training and i dont have to process the train_mask explicilty to ensure that only those nodes are trained?

**Reply by rusty1s** [reply]
At some point you will have to use `train_mask`, similar to how it is used in our [semi-supervised GCN example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/gcn.py#L59). PyG does not do any magic for you :)

**Reply by RamachandranSitaraman-hv** [reply]
In the example you have pointed to, the mask is used only for calculating loss, pred and acc after training, so if i want to limit the training itself only to those nodes defined by the train_mask i would have to implement that myself in the forward function, is that correct understanding? thanks

**Reply by rusty1s** [reply]
Yes, you can use `data.subgraph(data.train_mask)` for this.

**Reply by RamachandranSitaraman-hv** [reply]
Thats great, Could you please point me to an example code illustrating how data.subgraph is used...

**Reply by rusty1s** [reply]
I don't think we have an example of this. It will simply return a new `data` object (holding a subset of the data according to `train_mask`) which you can use just as the original one for GNN training:
```python
train_data = data.subgraph(data.train_mask)
train_out = model(train_data.x, train_data.edge_index)
loss = loss_fn(train_out, train_data.y)
```
Let me know if something is not clear.

**Reply by RamachandranSitaraman-hv** [reply]
This is very clear, thank you
